pyiron_contrib/tinybase/storage.py

```diff
@@ -0,0 +1,264 @@
+        """
+        pass
+
+    # DESIGN: this mostly exists to help to_object()ing GenericTinyJob, but it introduces a circular-ish connection.
+    # Maybe there's another way to do it?
+    @property
+    @abc.abstractmethod
+    def project(self):
+        """
+        The project that this storage belongs to.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def name(self):
+        """
+        The name of the group that this object is currently pointing to.
+
+        (In ProjectHDFio speak, this is the os.path.basename of h5_path)
+        """
+        pass
+
+    def to_object(self) -> "Storable":
+        """
+        Instantiate an object serialized to this group.
+
+        Returns:
+            :class:`.Storable`: deserialized object
+
+        Raises:
+            ValueError: no was object serialized in this group (either NAME, MODULE or VERSION values missing)
+            RuntimeError: failed to import serialized object (probably some libraries not installed)
+        """
+        try:
+            name = self["NAME"]
+            module = self["MODULE"]
+            version = self["VERSION"]
+        except KeyError:
+            raise ValueError("No object was serialized in this group!")
+        try:
+            cls = getattr(importlib.import_module(module), name)
+        except ImportError:
+            raise RuntimeError("Failed to import serialized object!")
+        return cls.restore(self, version=version)
+
+class ProjectHDFioStorageAdapter(GenericStorage):
+    """
+    Adapter class around ProjectHDFio to let it be used as a GenericStorage.
+    """
+
+    def __init__(self, project, hdf):
+        self._project = project
+        self._hdf = hdf
+
+    def __getitem__(self, item):
+        return self._hdf[item]
+
+    def __setitem__(self, item, value):
+        self._hdf[item] = value
+
+    def create_group(self, name):
+        return ProjectHDFioStorageAdapter(self._project, self._hdf.create_group(name))
+
+    def list_nodes(self):
+        return self._hdf.list_nodes()
+
+    def list_groups(self):
+        return self._hdf.list_groups()
+
+    # compat with small bug in base ProjectHDFio
+    list_dirs = list_groups
+
+    @property
+    def project(self):
+        return self._project
+
+    @property
+    def name(self):
+        return self._hdf.name
+
+# DESIGN: equivalent of HasHDF but with generalized language
+class Storable(abc.ABC):
+    """
+    Interface for classes that can store themselves to a :class:`~.GenericStorage`
+
+    Necessary overrides are :meth:`._store` and :meth:`._restore`.
+    """
+
+    @abc.abstractmethod
+    def _store(self, storage):
+        pass
+
+    def _store_type(self, storage):
+        storage["NAME"] = self.__class__.__name__
+        storage["MODULE"] = self.__class__.__module__
+        # DESIGN: what happens with objects defined in different parts of pyiron{_contrib,_atomistics,_base}?  Which
+        # version do we save?
```

Comment:
If people are using the conda installs, storing only the module that actually owns the class should be fine, since we always hard-limit our version numbers on the dependencies.

If it's not a version but some random commit, and they're working with bleeding edge stuff that is somehow interdependent, then I'm not so sure what will happen...